src/agent/protocols/protos/image.proto

```diff
@@ -0,0 +1,31 @@
+//
+// Copyright (c) 2021 Alibaba Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+syntax = "proto3";
+
+option go_package = "github.com/kata-containers/kata-containers/src/runtime/virtcontainers/pkg/agent/protocols/grpc";
+
+package grpc;
+
+// Image defines the public APIs for managing images.
+service Image {
+    // PullImage pulls an image with authentication config.
+    rpc PullImage(PullImageRequest) returns (PullImageResponse) {}
+}
+
+message PullImageRequest {
+    // Image name (e.g. docker.io/library/busybox:latest).
+    string image = 1;
+    // Unique image identifier, used to avoid duplication when unpacking the image layers.
+    string container_id = 2;
+    // Use USERNAME[:PASSWORD] for accessing the registry
+    string source_creds = 3;
```

Comment:
removed

Reply:
@sameo - what is the plan for getting source credentials for containers registries back in? Obviously we want to get them via the AA, but I'm not sure that is available yet. We agreed that this image service switch over wouldn't regress currently working function and without being about to pull images from authenticated registries that limits some of our ability to test.

Reply:
@stevenhorsman I believe there are at least a couple of opens related to that:

1. Is the registry authentication part of the CC trust boundary? The current way to have k8s pull an image from authenticated registries is to pass a secret through the pod description. Is that breaking the CC threat model? The host will be able to pull the CC container image because it knows the registry credentials, but the image will be encrypted (or signed). Having the registry credentials won't let the host tamper with or read the container image content, imho. So I wonder if it really makes sense to get those credentials only via attestation.
2.  **If** we want to support a model where one could retrieve registry authentication credentials through attestation, the current KBC and KBS solution for that are still TBD. Discussions are starting in the attestation-agent repo, but we don't have any clear answer for that atm.

Unless there's a strong disagreement about my first point, I think for now we could only focus on getting those credentials through the current k8s path. Given that our `source_creds` field of the `PullImage` payload would be unused in a k8s e2e context (the shim wont fill it, the AA won't neither), I proposed that we remove it for now and add authenticated registries back as its own standalone PR, after merging that one and #2733. This PR would add support for authenticated registries on top of a working k8s e2e flow for CC by:

1. Having containerd forward the `AuthConfig` information that it gets from CRI to the kata shim  
```grpc
// AuthConfig contains authorization information for connecting to a registry.
message AuthConfig {
    string username = 1;
    string password = 2;
    string auth = 3;
    string server_address = 4;
    // IdentityToken is used to authenticate the user and get
    // an access token for the registry.
    string identity_token = 5;
    // RegistryToken is a bearer token to be sent to a registry
    string registry_token = 6;
}
```
2. Defining our own `AuthConfig` subset and add it as one additional field to our agent `PullImage` API
3. Having the shim convert CRI's `AuthConfig` into our own and passing it as part of the agent `PullImage` request. 
4. Using that information to pass the credentials to `skopeo` (which is essentially the part that we have today).

With those 4 points we'd have something that could work e2e with k8s, for authenticated registries.

Let me know if that's an acceptable approach/plan or not. If it is, I'll create an issue to track that work so that we can push it as soon as #2956 and #2733 are in.

Reply:
Agree, let go along the first proposal until there's security threaten found.

Reply:
This makes sense.

Reply:
So I'm definitely not an expert, but my initial thoughts are:
- I don't know if passing keys through as secrets breaks the CC threat model, but I'm not sure it is great practice for what we are after. Unless we have an encryption scheme for those secrets then can't a CSP admin just spin up a different pod that will get list out the secrets and thereby get the credentials? If we ensure encryption, or verification for all images that it shouldn't effect the workload, but I wonder if there are things like DoS attacks (certain registries like docker hub have limits for number of images a user can pull), or other security implications of exposing the credentials?
- My expectation was that we wanted to get authentication credentials from the attestation agent's KBS, but as you say that isn't ready yet.
- I agree that the current solution is not the end one, but it wasn't intended to be. It was a temporary solution so we could test pulling a variety of images from authenticated registries in the kata agent, hence why I didn't extend that into the kata shim when adding ctr pull-image support. In the absence of generic key support from the attestation agent (which I think might cause other problems) I don't have an issue with your proposed implementation of building the model on the containerd AuthConfig either, but what I still don't understand is why we want to remove something that current works and we are using for testing without having a replacement. What was the gain from doing extra work to delete this code that I missed?

Reply:
>     * I don't know if passing keys through as secrets breaks the CC threat model, but I'm not sure it is great practice for what we are after. Unless we have an encryption scheme for those secrets then can't a CSP admin just spin up a different pod that will get list out the secrets and thereby get the credentials? If we ensure encryption, or verification for all images that it shouldn't effect the workload, but I wonder if there are things like DoS attacks (certain registries like docker hub have limits for number of images a user can pull), or other security implications of exposing the credentials?
> 

I agree, but I think those are all problems that exist today outside of the CC context. And what I'm saying is that the CC threat model is not covering this. Using authenticated registries is definitely a good practice, but I don't think it protects workload data in-use.

And to be clear, even with an in-guest image management model, we do want to eventually support authenticated registries.


>     * I agree that the current solution is not the end one, but it wasn't intended to be. It was a temporary solution so we could test pulling a variety of images from authenticated registries in the kata agent, hence why I didn't extend that into the kata shim when adding ctr pull-image support. 

Yes, that was clear to me as well.

> In the absence of generic key support from the attestation agent (which I think might cause other problems) I don't have an issue with your proposed implementation of building the model on the containerd AuthConfig either

Thanks.

> but what I still don't understand is why we want to remove something that current works and we are using for testing without having a replacement. What was the gain from doing extra work to delete this code that I missed?

You did not miss anything. It's just that with the next set of milestones this code would not be exercised except through `agent-ctl`. And it will be modified into something slightly more structured when we actually will support it end-to-end. I felt it was clearer to not continue supporting this when we're saying that our next goals are really focused on the k8s e2e pipeline.

If that's really problematic and slows our progress down, I can quickly add it back as a separate commit on top of that PR.

Reply:
Ok, thanks for the explanation. I don't want to hold up this PR, so I think the pragmatic path is to get this in as-is, but if the containerd k8s e2e solution isn't added soon then I can re-create the PR to temporary add agent-ctl source credentials support back in for temporary testing purposes.

Reply:
Hey @stevenhorsman I actually removed the commit that deleted `source_creds` and kept it as is for now.

Your request is fair, sorry for the back and forth. Do you think you could help with actually improving the current state and using a more structured agent API payload for registry authentication credentials, something that would be a subset of `AuthConfig` for now like e.g.

```
message AuthConfig {
    string username = 1;
    string password = 2;
}
```

Reply:
Sure, I think we should create an issue and get a bit of design done. I'm not sure it is as simple as username and password as some (most of the container registries we use recommend using API keys and some of these have a pre-prepended userid and other don't. That's why we originally kept it as a single combined string to push that complication on the user, but understanding what the kubernetes secret flows down into the containerd `AuthConfig` would be really useful. I think as get closer to moving to the Alibaba containerd fork we might learn a bit more.

Reply:
Right. With a simple username+password `AuthConfig` we know we're not overengineering it at least, and we'd get the same level of functionality as we have today. And we can expand it with what we need as we learn more, without breaking much.